index.js

```diff
@@ -143,42 +145,66 @@ module.exports = {
       }
 
-      get(groupId, (err, { writeKey, root }) => {
+      get(groupId, (err, { root }) => {
         // prettier-ignore
         if (err) return cb(clarify(err, `Failed to get group details when adding members`))
 
         getRootFeedIdFromMsgId(root, (err, rootAuthorId) => {
           // prettier-ignore
           if (err) return cb(clarify(err, "couldn't get root id of author of root msg"))
 
-          const content = {
-            type: 'group/add-member',
-            version: 'v2',
-            groupKey: writeKey.key.toString('base64'),
-            root,
-            creator: rootAuthorId,
-            recps: [groupId, ...feedIds],
-          }
+          getTipEpochs(groupId, (err, tipEpochs) => {
+            // prettier-ignore
+            if (err) return cb(clarify(err, "Couldn't get tip epochs when adding members"))
 
-          if (opts.text) content.text = opts.text
+            const getFeed = opts?._feedKeys
+              ? (cb) => cb(null, { keys: opts._feedKeys })
+              : findOrCreateAdditionsFeed
 
-          const getFeed = opts?._feedKeys
-            ? (cb) => cb(null, { keys: opts._feedKeys })
-            : findOrCreateAdditionsFeed
+            getFeed((err, additionsFeed) => {
+              // prettier-ignore
+              if (err) return cb(clarify(err, 'Failed to find or create additions feed when adding members'))
 
-          getFeed((err, additionsFeed) => {
-            // prettier-ignore
-            if (err) return cb(clarify(err, 'Failed to find or create additions feed when adding members'))
+              const options = {
+                isValid: isAddMember,
+                tangles: ['members'],
+                feedKeys: additionsFeed.keys,
+              }
+              pull(
+                pull.values(tipEpochs),
+                pull.asyncMap((tipEpoch, cb) => {
+                  getPredecessorEpochs(
+                    groupId,
+                    tipEpoch.id,
+                    (err, predecessors) => {
+                      // prettier-ignore
+                      if (err) return cb(clarify(err, "Couldn't get predecessor epochs when adding members"))
+
+                      const oldSecrets = predecessors.map((pred) =>
+                        pred.secret.toString('base64')
+                      )
+                      const content = {
+                        type: 'group/add-member',
+                        version: 'v2',
+                        secret: tipEpoch.secret.toString('base64'),
+                        oldSecrets: opts.oldSecrets ? oldSecrets : undefined,
+                        root,
+                        creator: rootAuthorId,
+                        text: opts?.text,
+                        recps: [groupId, ...feedIds],
+                      }
+                      return cb(null, content)
+                    }
+                  )
+                }),
+                pull.asyncMap((content, cb) => publish(content, options, cb)),
```

Comment:
I think our publish method should watch the types coming through and if there's a group-relates message then do schema validations before publish. Different PR though

Reply:
`tribes2.publish` does schema validations though?

Reply:
oh i forgot that was done via opts.
Bravo!